pkg/controlplane/controller/kubernetesservice/controller.go

```diff
@@ -67,17 +73,23 @@ type Config struct {
 }
 
 // New returns a controller for watching the kubernetes service endpoints.
-func New(config Config, client kubernetes.Interface) *Controller {
+func New(config Config, client kubernetes.Interface, serviceInformer v1informers.ServiceInformer) *Controller {
 	return &Controller{
-		Config: config,
-		client: client,
-		stopCh: make(chan struct{}),
+		Config:        config,
+		client:        client,
+		serviceLister: serviceInformer.Lister(),
+		serviceSynced: serviceInformer.Informer().HasSynced,
+		stopCh:        make(chan struct{}),
 	}
 }
 
 // Start begins the core controller loops that must exist for bootstrapping
 // a cluster.
 func (c *Controller) Start(stopCh <-chan struct{}) {
+	if !cache.WaitForCacheSync(stopCh, c.serviceSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
```

Comment:
This slows down when we create, update or delete endpoints. Just worth another thought whether we could imagine cases where we cannot sync services yet, but need to update the service quickly.

Reply:
The service should only publish the endpoints when it is ready, the Run loop already waits for the `/readyz` endpoint to return OK, so it does not really matter if we wait here too, because we'll have to wait later, is just this looks more "correct"

Reply:
see line 146

```
	
func (c *Controller) Run(ch <-chan struct{}) {
	// wait until process is ready
	wait.PollImmediateUntil(100*time.Millisecond, func() (bool, error) {
		var code int
		c.client.CoreV1().RESTClient().Get().AbsPath("/readyz").Do(context.TODO()).StatusCode(&code)
		return code == http.StatusOK, nil
	}, ch)
```

Reply:
necessary condition vs sufficient condition thing, this is necessary but not sufficient XD

Reply:
+1